rnd/autogpt_server/autogpt_server/executor/manager.py

```diff
@@ -350,76 +354,111 @@ def on_node_execution(cls, q: ExecutionQueue[NodeExecution], data: NodeExecution
+                running_executions[exec_data.node_id] = cls.executor.apply_async(
+                    cls.on_node_execution,
+                    (queue, exec_data),
+                    callback=lambda _: running_executions.pop(exec_data.node_id),
                 )
 
                 # Avoid terminating graph execution when some nodes are still running.
-                while queue.empty() and futures:
-                    for node_id, future in list(futures.items()):
-                        if future.done():
-                            del futures[node_id]
-                        elif queue.empty():
-                            cls.wait_future(future)
+                while queue.empty() and running_executions:
+                    for execution in list(running_executions.values()):
+                        if cancel.is_set():
+                            return
+
+                        if not queue.empty():
+                            break  # yield to parent loop to execute new queue items
+
+                        execution.wait(3)
 
             logger.warning(f"{prefix} Finished graph execution")
         except Exception as e:
             logger.exception(f"{prefix} Failed graph execution: {e}")
-
-    @classmethod
-    def wait_future(cls, future: Future, timeout: int | None = 3):
-        try:
-            if not future.done():
-                future.result(timeout=timeout)
-        except TimeoutError:
-            # Avoid being blocked by long-running node, by not waiting its completion.
-            pass
+        finally:
+            if not cancel.is_set():
+                finished = True
+                cancel.set()
+            cancel_thread.join()
 
 
 class ExecutionManager(AppService):
     def __init__(self):
         self.pool_size = Config().num_graph_workers
         self.queue = ExecutionQueue[GraphExecution]()
+        self.active_graph_runs: dict[str, tuple[Future, threading.Event]] = {}
 
     def run_service(self):
         with ProcessPoolExecutor(
             max_workers=self.pool_size,
             initializer=Executor.on_graph_executor_start,
         ) as executor:
-            logger.warning(
+            sync_manager = multiprocessing.Manager()
+            logger.info(
                 f"Execution manager started with max-{self.pool_size} graph workers."
             )
             while True:
-                executor.submit(Executor.on_graph_execution, self.queue.get())
+                graph_exec_data = self.queue.get()
+                graph_exec_id = graph_exec_data.graph_exec_id
+                cancel_event = sync_manager.Event()
+                future = executor.submit(
+                    Executor.on_graph_execution, graph_exec_data, cancel_event
+                )
+                self.active_graph_runs[graph_exec_id] = (future, cancel_event)
+                future.add_done_callback(
+                    lambda _: self.active_graph_runs.pop(graph_exec_id)
```

Comment:
Should we just do this on the wrapped try/finally block ?
Otherwise it will never pop on failure

Reply:
It should never fail, since the entire `on_graph_execution` is wrapped in a try/except/finally, right?

I also don't see how we could manipulate the `self.active_graph_runs` object from `on_graph_execution`, since it runs in a different process and contains a reference to the `Future` which probably isn't picklable.

Reply:
I see, yeah good point